src/wasm/wasm-ir-builder.cpp

```diff
@@ -759,13 +759,23 @@ Result<Expression*> IRBuilder::finishScope(Block* block) {
   auto type = scope.getResultType();
 
   if (scope.unreachable) {
-    // Drop everything before the last unreachable.
+    // Drop everything before the last unreachable expression.
     bool sawUnreachable = false;
     for (int i = scope.exprStack.size() - 1; i >= 0; --i) {
       if (sawUnreachable) {
         scope.exprStack[i] = builder.dropIfConcretelyTyped(scope.exprStack[i]);
       } else if (scope.exprStack[i]->type == Type::unreachable) {
         sawUnreachable = true;
+      } else {
+        // As a special case, look for unreachable if conditions, since they may
+        // not make their ifs unreachable.
+        for (auto* curr = scope.exprStack[i]->dynCast<If>(); curr;
+             curr = curr->condition->dynCast<If>()) {
+          if (curr->condition->type == Type::unreachable) {
+            sawUnreachable = true;
+            break;
+          }
+        }
```

Comment:
Would it be simpler to change this?

https://github.com/WebAssembly/binaryen/blob/e13bf0fb72fca160f457570b930c4ba3c35ead3a/src/wasm/wasm.cpp#L230

I think we could just make an if with an unreachable condition be unreachable, unless I'm missing a reason not to.

Reply:
(that would avoid special-casing here, and seems simpler overall: the code is dead, so let's mark it dead)